File: functions/approveProcessedSensorV2/s3UpdaterV2.js
import { S3Client, GetObjectCommand, PutObjectCommand } from '@aws-sdk/client-s3';

const s3Client = new S3Client(
  process.env.IS_LOCAL
    ? {
        region: 'us-east-2',
        endpoint: 'http://host.docker.internal:9090',
        forcePathStyle: true,
        credentials: { accessKeyId: 'test', secretAccessKey: 'test' },
      }
    : {
        region: 'auto',
        endpoint: process.env.R2_ENDPOINT,
        credentials: {
          accessKeyId: process.env.R2_ACCESS_KEY_ID,
          secretAccessKey: process.env.R2_SECRET_ACCESS_KEY,
        },
      }
);

const BUCKET = process.env.IS_LOCAL
  ? (process.env.S3_BUCKET_NAME || 'my-test-bucket')
  : process.env.R2_BUCKET_NAME;

const streamToBuffer = (stream) =>
  new Promise((resolve, reject) => {
    const chunks = [];
    stream.on('data', (c) => chunks.push(c));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
    stream.on('error', reject);
  });

const getJson = async (key) => {
  const res = await s3Client.send(new GetObjectCommand({ Bucket: BUCKET, Key: key }));
  const buf = await streamToBuffer(res.Body);
  return JSON.parse(buf.toString());
};

const putJson = (key, obj) =>
  s3Client.send(new PutObjectCommand({
    Bucket: BUCKET,
    Key: key,
    Body: JSON.stringify(obj, null, 2),
    ContentType: 'application/json',
  }));

const isNotFound = (err) => err?.name === 'NoSuchKey' || err?.Code === 'NoSuchKey' || err?.$metadata?.httpStatusCode === 404;

// V2 published statics live under the `v2/` key prefix on R2 (the FE reads
// https://groov-api.com/v2/index.json etc.). The bucket root holds the V1
// files, so every V2 key must carry this prefix.
const V2_PREFIX = 'v2/';

// Walk V2 sensor proteins and collect unique ligand names.
const collectLigandNames = (data) => {
  const names = new Set();
  for (const protein of data.proteins ?? []) {
    for (const stim of protein.stimulus ?? []) {
      // Tolerate both stimulusType (camelCase, written by addNewSensorV2) and stimulus_type (snake_case, migrated data)
      const types = stim.stimulusType ?? stim.stimulus_type ?? [];
      for (const t of types) {
        for (const m of t.small_molecule ?? []) {
          if (m?.name) names.add(m.name);
        }
      }
    }
  }
  return [...names];
};

const buildIndexEntry = (data, category, grv_id) => {
  const first = data.proteins?.[0] ?? {};
  return {
    id: grv_id,
    alias: first.alias ?? '',
    uniprot_id: first.uniprot_id ?? '',
    organism_name: first.origin?.[0]?.organism_name ?? '',
    category,
    ligands: collectLigandNames(data),
  };
};

const buildFamilyIndexEntry = (data, grv_id) => {
  const first = data.proteins?.[0] ?? {};
  return {
    id: grv_id,
    alias: first.alias ?? '',
    uniprot_id: first.uniprot_id ?? '',
    kegg_id: first.kegg_id ?? null,
    organism_name: first.origin?.[0]?.organism_name ?? '',
    ligands: collectLigandNames(data),
  };
};

const updateMainIndex = async (data, category, grv_id) => {
  let index;
  try {
    index = await getJson(`${V2_PREFIX}index.json`);
  } catch (err) {
    if (!isNotFound(err)) throw err;
    index = { stats: { regulators: 0, ligands: 0 }, sensors: [] };
  }
  if (!Array.isArray(index.sensors)) index.sensors = [];
  const entry = buildIndexEntry(data, category, grv_id);
  const existing = index.sensors.findIndex((s) => s.id === grv_id);
  if (existing >= 0) index.sensors[existing] = entry;
  else index.sensors.push(entry);

  const allLigands = new Set();
  index.sensors.forEach((s) => (s.ligands ?? []).forEach((l) => allLigands.add(l)));
  index.stats = { regulators: index.sensors.length, ligands: allLigands.size };

  await putJson(`${V2_PREFIX}index.json`, index);
};

const updateFamilyIndex = async (data, category, grv_id) => {
  const key = `${V2_PREFIX}indexes/${category.toLowerCase()}.json`;
  let familyIndex;
  try {
    familyIndex = await getJson(key);
  } catch (err) {
    if (!isNotFound(err)) throw err;
    familyIndex = { count: 0, data: [] };
  }
  const entry = buildFamilyIndexEntry(data, grv_id);
  const existing = familyIndex.data.findIndex((s) => s.id === grv_id);
  if (existing >= 0) familyIndex.data[existing] = entry;
  else familyIndex.data.push(entry);
  familyIndex.count = familyIndex.data.length;
  await putJson(key, familyIndex);
};

const saveSensorFile = async (data, category, grv_id) => {
  await putJson(`${V2_PREFIX}sensors/${category.toLowerCase()}/${grv_id}.json`, data);
};

const updateAllSensors = async (data) => {
  let all;
  try {
    all = await getJson(`${V2_PREFIX}all-sensors.json`);
  } catch (err) {
    if (!isNotFound(err)) throw err;
    all = { version: new Date().toISOString(), count: 0, sensors: [] };
  }
  const existing = all.sensors.findIndex((s) => s.id === data.id);
  if (existing >= 0) all.sensors[existing] = data;
  else all.sensors.push(data);
  all.count = all.sensors.length;
  all.version = new Date().toISOString();
  await putJson(`${V2_PREFIX}all-sensors.json`, all);
};

const zeroPad = (n, w) => String(n).padStart(w, '0');

// Returns the next GRV-ID for the given single-character prefix by scanning R2 index.json.
// Per-prefix counters live implicitly in the index — one prefix per category for single-component
// sensors, plus prefix 'D' shared by all two-component sensors.
export const mintNextGrvId = async (prefix) => {
  let index;
  try {
    index = await getJson(`${V2_PREFIX}index.json`);
  } catch (err) {
    if (!isNotFound(err)) throw err;
    index = { sensors: [] };
  }
  const pat = new RegExp(`^GRV-${prefix}(\\d{5})$`);
  let max = 0;
  for (const s of index.sensors ?? []) {
    const m = s?.id ? pat.exec(s.id) : null;
    if (m) {
      const n = parseInt(m[1], 10);
      if (n > max) max = n;
    }
  }
  return `GRV-${prefix}${zeroPad(max + 1, 5)}`;
};

export const regenerateStaticJSON = async (data, category, grv_id) => {
  await updateMainIndex(data, category, grv_id);
  await updateFamilyIndex(data, category, grv_id);
  await saveSensorFile(data, category, grv_id);
  await updateAllSensors(data);
};

File: functions/approveProcessedSensorV2/s3_updater_v2.py
import datetime
import json
import os
import re
from decimal import Decimal

import boto3
import botocore.exceptions

# V2 published statics live under the `v2/` key prefix on R2 (the FE reads
# https://groov-api.com/v2/index.json etc.). The bucket root holds the V1
# files, so every V2 key must carry this prefix.
V2_PREFIX = "v2/"


def _s3_client():
    if os.environ.get("IS_LOCAL"):
        # No endpoint_url override: boto3 auto-targets the Floci-injected
        # AWS_ENDPOINT_URL env var, and credentials come from the Lambda
        # runtime's own credential provider (also Floci-injected) — same
        # pattern as the DynamoDB clients in this codebase.
        return boto3.client("s3", region_name="us-east-2")
    return boto3.client(
        "s3",
        region_name="auto",
        endpoint_url=os.environ.get("R2_ENDPOINT"),
        aws_access_key_id=os.environ.get("R2_ACCESS_KEY_ID"),
        aws_secret_access_key=os.environ.get("R2_SECRET_ACCESS_KEY"),
    )


def _bucket():
    if os.environ.get("IS_LOCAL"):
        return os.environ.get("S3_BUCKET_NAME") or "my-test-bucket"
    return os.environ.get("R2_BUCKET_NAME")


def _coalesce(*values):
    """Nullish-coalescing helper: first argument that is not None, else None."""
    for v in values:
        if v is not None:
            return v
    return None


def _get_json(key):
    client = _s3_client()
    resp = client.get_object(Bucket=_bucket(), Key=key)
    return json.loads(resp["Body"].read())


def _json_default(value):
    # The published statics are built from the sensor `data` read out of
    # DynamoDB, whose numbers are Decimal — json.dumps can't serialize those.
    # Emit integral values as int and the rest as float, matching JS
    # JSON.stringify so the FE sees plain numbers.
    if isinstance(value, Decimal):
        return int(value) if value % 1 == 0 else float(value)
    raise TypeError(f"Object of type {type(value).__name__} is not JSON serializable")


def _put_json(key, obj):
    client = _s3_client()
    client.put_object(
        Bucket=_bucket(),
        Key=key,
        Body=json.dumps(obj, indent=2, default=_json_default),
        ContentType="application/json",
    )


def _is_not_found(err):
    if not isinstance(err, botocore.exceptions.ClientError):
        return False
    code = err.response.get("Error", {}).get("Code")
    if code in ("NoSuchKey", "404"):
        return True
    return err.response.get("ResponseMetadata", {}).get("HTTPStatusCode") == 404


# Walk V2 sensor proteins and collect unique ligand names (insertion order,
# mirroring the JS Set -> [...names] behavior).
def _collect_ligand_names(data):
    names = []
    seen = set()
    for protein in data.get("proteins") or []:
        for stim in protein.get("stimulus") or []:
            # Tolerate both stimulusType (camelCase, written by addNewSensorV2)
            # and stimulus_type (snake_case, migrated data).
            types = _coalesce(stim.get("stimulusType"), stim.get("stimulus_type"), [])
            for t in types:
                for m in t.get("small_molecule") or []:
                    name = (m or {}).get("name")
                    if name and name not in seen:
                        seen.add(name)
                        names.append(name)
    return names


def _build_index_entry(data, category, grv_id):
    proteins = data.get("proteins") or []
    first = proteins[0] if proteins else {}
    origin = first.get("origin") or []
    organism_name = _coalesce(origin[0].get("organism_name") if origin else None, "")
    return {
        "id": grv_id,
        "alias": _coalesce(first.get("alias"), ""),
        "uniprot_id": _coalesce(first.get("uniprot_id"), ""),
        "organism_name": organism_name,
        "category": category,
        "ligands": _collect_ligand_names(data),
    }


def _build_family_index_entry(data, grv_id):
    proteins = data.get("proteins") or []
    first = proteins[0] if proteins else {}
    origin = first.get("origin") or []
    organism_name = _coalesce(origin[0].get("organism_name") if origin else None, "")
    return {
        "id": grv_id,
        "alias": _coalesce(first.get("alias"), ""),
        "uniprot_id": _coalesce(first.get("uniprot_id"), ""),
        "kegg_id": _coalesce(first.get("kegg_id"), None),
        "organism_name": organism_name,
        "ligands": _collect_ligand_names(data),
    }


def _update_main_index(data, category, grv_id):
    try:
        index = _get_json(f"{V2_PREFIX}index.json")
    except Exception as err:
        if not _is_not_found(err):
            raise
        index = {"stats": {"regulators": 0, "ligands": 0}, "sensors": []}
    if not isinstance(index.get("sensors"), list):
        index["sensors"] = []
    sensors = index["sensors"]
    entry = _build_index_entry(data, category, grv_id)
    existing_idx = next((i for i, s in enumerate(sensors) if s.get("id") == grv_id), -1)
    if existing_idx >= 0:
        sensors[existing_idx] = entry
    else:
        sensors.append(entry)

    all_ligands = set()
    for s in sensors:
        for ligand in s.get("ligands") or []:
            all_ligands.add(ligand)
    index["stats"] = {"regulators": len(sensors), "ligands": len(all_ligands)}

    _put_json(f"{V2_PREFIX}index.json", index)


def _update_family_index(data, category, grv_id):
    key = f"{V2_PREFIX}indexes/{category.lower()}.json"
    try:
        family_index = _get_json(key)
    except Exception as err:
        if not _is_not_found(err):
            raise
        family_index = {"count": 0, "data": []}
    items = family_index.get("data") or []
    entry = _build_family_index_entry(data, grv_id)
    existing_idx = next((i for i, s in enumerate(items) if s.get("id") == grv_id), -1)
    if existing_idx >= 0:
        items[existing_idx] = entry
    else:
        items.append(entry)
    family_index["data"] = items
    family_index["count"] = len(items)
    _put_json(key, family_index)


def _save_sensor_file(data, category, grv_id):
    _put_json(f"{V2_PREFIX}sensors/{category.lower()}/{grv_id}.json", data)


def _update_all_sensors(data):
    try:
        all_sensors = _get_json(f"{V2_PREFIX}all-sensors.json")
    except Exception as err:
        if not _is_not_found(err):
            raise
        all_sensors = {
            "version": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            "count": 0,
            "sensors": [],
        }
    sensors = all_sensors.get("sensors") or []
    existing_idx = next((i for i, s in enumerate(sensors) if s.get("id") == data.get("id")), -1)
    if existing_idx >= 0:
        sensors[existing_idx] = data
    else:
        sensors.append(data)
    all_sensors["sensors"] = sensors
    all_sensors["count"] = len(sensors)
    all_sensors["version"] = datetime.datetime.now(datetime.timezone.utc).isoformat()
    _put_json(f"{V2_PREFIX}all-sensors.json", all_sensors)


def _zero_pad(n, width):
    return str(n).zfill(width)


# Returns the next GRV-ID for the given single-character prefix by scanning R2 index.json.
# Per-prefix counters live implicitly in the index — one prefix per category for single-component
# sensors, plus prefix 'D' shared by all two-component sensors.
def mint_next_grv_id(prefix):
    try:
        index = _get_json(f"{V2_PREFIX}index.json")
    except Exception as err:
        if not _is_not_found(err):
            raise
        index = {"sensors": []}
    pattern = re.compile(rf"^GRV-{re.escape(prefix)}(\d{{5}})$")
    max_n = 0
    for s in index.get("sensors") or []:
        sid = s.get("id") if s else None
        if sid:
            m = pattern.match(sid)
            if m:
                n = int(m.group(1))
                if n > max_n:
                    max_n = n
    return f"GRV-{prefix}{_zero_pad(max_n + 1, 5)}"


def regenerate_static_json(data, category, grv_id):
    _update_main_index(data, category, grv_id)
    _update_family_index(data, category, grv_id)
    _save_sensor_file(data, category, grv_id)
    _update_all_sensors(data)
